### ISSUE TITLE
Cancelled error on invalid index entries

### ISSUE THREAD

**Issue by jonhoo** [Byron/crates-index-diff-rs#35]
`crates-index-diff` now yields `DiffForEach(Diff(Cancelled))` error on seemingly-okay git history (version 12 worked fine):

```console
$ git init repo
$ mkdir -p repo/aw/s-
$ echo '{"name":"aws-foo","vers":"0.0.1","deps":[],"cksum":"","features":{},"yanked":false}' > repo/aw/s-/aws-foo
$ git -C repo add .
$ git -C repo commit -m 'commit'
# note down the commit hash here
$ echo '{"name":"aws-foo","vers":"0.0.2","deps":[],"cksum":"","features":{},"yanked":false}' >> repo/aw/s-/aws-foo
$ git -C repo add .
$ git -C repo commit -m 'commit'
# note down this commit hash too
$ git -C repo remote add origin origin
$ git -C repo diff <first commit hash>..<second commit hash>
diff --git a/aw/s-/aws-foo b/aw/s-/aws-foo
index f512d82..9bff708 100644
--- a/aw/s-/aws-foo
+++ b/aw/s-/aws-foo
@@ -1 +1,2 @@
 {"name":"aws-foo","vers":"0.0.1","deps":[],"cksum":"","features":{},"yanked":false}
+{"name":"aws-foo","vers":"0.0.2","deps":[],"cksum":"","features":{},"yanked":false}
$ cargo new --bin whatever && cd whatever
$ echo 'crates-index-diff = "15"' >> Cargo.toml
$ cat src/main.rs
use crates_index_diff::{git, index::CloneOptions, Index};
use std::sync::atomic::AtomicBool;

fn main() {
    let repo = git::open("../repo").unwrap();
    let origin_url = repo
        .find_remote("origin")
        .unwrap()
        .url(git::remote::Direction::Fetch)
        .unwrap()
        .to_bstring()
        .to_string();
    let index = Index::from_path_or_cloned_with_options(
        "../repo",
        git::progress::Discard,
        &AtomicBool::default(),
        CloneOptions {
            url: origin_url,
            ..Default::default()
        },
    )
    .unwrap();

    index.changes("<first commit hash>", "<second commit hash>").unwrap();
}
$ cargo r
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: DiffForEach(Diff(Cancelled))', src/main.rs:24:41
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

The `origin_url` bit is to ensure we always use the existing path rather than ever try to clone. At least that was the intent, and that worked back in version 12. Don't know if it's relevant here.

**Comment by jonhoo** [comment]
I _think_ this is perhaps related to the index entries being malformed (e.g., `cksum` is `""`) — this was in a test — though the error doesn't indicate that.

**Comment by jonhoo** [comment]
Though even if I use a valid checksum like `da39a3ee5e6b4b0d3255bfef95601890afd80709` (the empty SHA1 hash), I get the same error.

**Comment by jonhoo** [comment]
Ah, SHA256, not SHA1. If I substitute `0000000000000000000000000000000000000000000000000000000000000000` as `cksum` in the entries this code runs fine. I think this then comes down to a better error message.

**Comment by Byron** [comment]
Thanks for letting me know and digging into possible causes - this helped tremendously to quickly validate that indeed it swallows an error where it shouldn't. This requires a fix in `gitoxide` and as `main`  is probably far enough to be another major release, it will take a moment to land here.

Maybe in the meantime, you can use a local version and [unwrap() here](https://github.com/Byron/crates-index-diff-rs/blob/44a5579feb424abcb2d6d75e0631577f1c71b33b/src/index/diff/mod.rs#L235) to see exactly what the problem is.

However, I'd expect that [it can't decode sha256](https://github.com/Byron/crates-index-diff-rs/blob/b42ac1e055bb8264804c40a2fe436e45850e9422/src/types.rs#L135-L136) and with the test above I should be reproducing it in no time so a fix should be possible without fixing `gitoxide` first.

**Comment by jonhoo** [comment]
All good — I just updated the test to include a dummy checksum, so this doesn't block me :)

**Comment by Byron** [comment]
Is SHA256 something to expect there now? I aligned the representation of versions with `crates-index` and it only knows SHA1 as well.

**Comment by jonhoo** [comment]
No, I also thought it had to be SHA1, but putting a SHA1 hash hex in the index entry produced the same error. Only with the longer hash I gave above did I avoid the error. Which does seem weird indeed.

**Comment by Byron** [comment]
Sorry for the confusion - it _has_ to be SHA256, and I believe that's what `crates-index` would also require or else fail to iterate the crates-index. Previously it worked (here) as it would keep the checksums as strings, but I recently optimized the representation to be more efficient.

**Comment by Byron** [comment]
My plan is to close this issue once there is a new release of `crates-index-diff` with the latest version of `git-repository` so that errors are properly communicated. Thus there is no plan to support other hashes, particularly shorter ones. Your input would be appreciated if anything is missing from that response.

**Comment by jonhoo** [comment]
No, that seems reasonable! All I was missing in this case was an error that pointed me to the bad hash — that would have made this trivial to fix.
